programs/validator-history/src/instructions/copy_cluster_info.rs

```diff
@@ -36,26 +39,64 @@ pub fn handle_copy_cluster_info(ctx: Context<CopyClusterInfo>) -> Result<()> {
 
     let epoch_start_timestamp = cast_epoch_start_timestamp(clock.epoch_start_timestamp);
 
-    // Sets the slot history for the previous epoch, since the current epoch is not yet complete.
-    if epoch > 0 {
-        cluster_history_account
-            .set_blocks(epoch - 1, blocks_in_epoch(epoch - 1, &slot_history)?)?;
-    }
-    cluster_history_account.set_blocks(epoch, blocks_in_epoch(epoch, &slot_history)?)?;
+    let epoch_schedule = EpochSchedule::get()?;
+
+    let slot_history = if epoch > 0 {
+        let slot_history_next_slot = slot_history.next_slot;
+        let start_slot = epoch_schedule.get_first_slot_in_epoch((epoch - 1).into());
+        let end_slot = epoch_schedule.get_last_slot_in_epoch((epoch - 1).into());
+        let (num_blocks, bitvec_inner) =
+            confirmed_blocks_in_epoch(start_slot, end_slot, *slot_history)?;
+        // Sets the slot history for the previous epoch, since the total number of blocks in the epoch is now final
+        cluster_history_account.set_blocks(epoch - 1, num_blocks)?;
+        // recreates SlotHistory with same heap memory chunk
+        Box::new(SlotHistory {
+            bits: bitvec_inner.into(),
```

Comment:
shouldn't this be for the new epoch? weird bitvec_inner generated by start_slot, end_slot here from previous epoch then we use it below fo the current epoch

Reply:
SlotHistory contains last 1,048,576 slots so it will always contain all slots from last epoch and this epoch. Recreating it here to avoid copying + ownership issues. Can add a comment about how it contains both epochs of slots